### PR TITLE
[CUDA][Kernel] A bunch of int64 kernels for COO and CSR

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

New kernels added for int64 CSR/COO on CUDA:
- [x] COOSort
- [x] CSRSort
- [x] COO2CSR
- [x] CSR2COO
- [x] CSR2CSC

TODO
~~* Integrate with python and fix bugs.~~

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: src/array/cpu/array_sort.cc [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+template <DLDeviceType XPU, typename IdType>
+std::pair<IdArray, IdArray> Sort(IdArray array) {
+  const int64_t nitem = array->shape[0];
+  IdArray val = array.Clone();
+  IdArray idx = aten::Range(0, nitem, 64, array->ctx);
+  IdType* val_data = val.Ptr<IdType>();
+  int64_t* idx_data = idx.Ptr<int64_t>();
+  typedef std::pair<IdType, int64_t> Pair;
+#ifdef PARALLEL_ALGORITHMS
```
Comment (yzh119) [inline src/array/cpu/array_sort.cc:166]: @classicsong , when would this flag be activated?

File: src/array/cuda/csr2coo.cu [hunk 2]
```diff
@@ -41,12 +47,67 @@ COOMatrix CSRToCOO(CSRMatrix csr) {
                    true, csr.sorted);
 }
 
+/*!
+ * \brief Repeat elements
+ *
+ * For example:
+ * val = [3, 0, 1]
+ * repeats = [1, 0, 2]
+ * pos = [0, 1, 1]
```
Comment (yzh119) [inline src/array/cuda/csr2coo.cu:56]: describe what's the purpose of pos?

File: src/array/cuda/csr2coo.cu [hunk 3]
```diff
@@ -85,6 +146,26 @@ COOMatrix CSRToCOODataAsOrder(CSRMatrix csr) {
   // The row and column field have already been reordered according
   // to data, thus the data field will be deprecated.
   coo.data = aten::NullArray();
+  coo.row_sorted = false;
```
Comment (yzh119) [inline src/array/cuda/csr2coo.cu:149]: Why `row_sorted` and `col_sorted` were set to false when calling CSRToCOODataAsOrder?

File: src/graph/heterograph.h [hunk 4]
```diff
@@ -15,6 +15,7 @@
 #include <vector>
 #include <set>
 #include <tuple>
+#include <memory>
```
Comment (yzh119) [inline src/graph/heterograph.h:18]: Is this change intentional?

File: src/array/cpu/array_sort.cc [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    return ret;
+  }
+
+  PairIterator operator-(const std::ptrdiff_t& movement) const {
+    PairIterator ret(*this);
+    ret -= movement;
+    return ret;
+  }
+
+  std::ptrdiff_t operator-(const PairIterator& other) const {
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply src/array/cpu/array_sort.cc:136]: It's part of the [random-access iterator interface](http://www.cplusplus.com/reference/iterator/). I guess some sort algorithms might use this operator.

File: src/graph/heterograph.h (hunk 4 above)
Comment (jermainewang) [reply src/graph/heterograph.h:18]: Yes, for fixing lint errors.

File: include/dgl/aten/array_ops.h [hunk 6]
```diff
@@ -304,6 +307,17 @@ IdArray CumSum(IdArray array, bool prepend_zero = false);
  */
 IdArray NonZero(NDArray array);
 
+/*!
+ * \brief Sort the ID vector in ascending order.
+ *
+ * It performs both sort and arg_sort (returning the sorted index). The sorted index
+ * is always in int64.
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply include/dgl/aten/array_ops.h:314]: I think int64 is the safest since the tensor shape is in int64.

File: src/array/cuda/csr2coo.cu (hunk 2 above)
Comment (jermainewang) [reply src/array/cuda/csr2coo.cu:56]: done
